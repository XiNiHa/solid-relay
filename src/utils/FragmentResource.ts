import {
  ID_KEY,
  RelayFeatureFlags,
  createOperationDescriptor,
  getFragmentIdentifier,
  getSelector,
  getVariablesFromFragment,
  isPromise,
  __internal,
  handlePotentialSnapshotErrors,
  recycleNodesInto,
  getPendingOperationsForFragment,
} from 'relay-runtime'
import type {
  ConcreteRequest,
  DataID,
  Disposable,
  IEnvironment,
  ReaderFragment,
  RequestDescriptor,
  Snapshot,
} from 'relay-runtime'
import { getPromiseForActiveRequest } from 'relay-runtime/lib/query/fetchQueryInternal'
import type { MissingLiveResolverField } from 'relay-runtime/lib/store/RelayStoreTypes'
import invariant from 'tiny-invariant'

import * as LRUCache from './LRUCache'
import {
  QueryResource,
  QueryResult,
  getQueryResourceForEnvironment,
} from './QueryResource'
import SuspenseResource from './SuspenseResource'

type FragmentResourceCache = LRUCache.Cache<
  | {
      kind: 'pending'
      pendingOperations: readonly RequestDescriptor[]
      promise: Promise<void>
      result: FragmentResult
    }
  | { kind: 'done'; result: FragmentResult }
>

const WEAKMAP_SUPPORTED = typeof WeakMap === 'function'
interface IMap<K, V> {
  get(key: K): V | void
  set(key: K, value: V): IMap<K, V>
}

type SingularOrPluralSnapshot = Snapshot | readonly Snapshot[]

export type FragmentResult = {
  cacheKey: string
  data: unknown
  isMissingData: boolean
  snapshot: SingularOrPluralSnapshot | null
  storeEpoch: number
}

const CACHE_CAPACITY = 1000000

const CONSTANT_READONLY_EMPTY_ARRAY = Object.freeze([])

function isMissingData(snapshot: SingularOrPluralSnapshot): boolean {
  if ('length' in snapshot) {
    return snapshot.some((s) => s.isMissingData)
  }
  return snapshot.isMissingData
}

function hasMissingClientEdges(snapshot: SingularOrPluralSnapshot): boolean {
  if ('length' in snapshot) {
    return snapshot.some((s) => (s.missingClientEdges?.length ?? 0) > 0)
  }
  return (snapshot.missingClientEdges?.length ?? 0) > 0
}

function missingLiveResolverFields(
  snapshot: SingularOrPluralSnapshot
): readonly MissingLiveResolverField[] | undefined {
  if ('length' in snapshot) {
    return snapshot
      .map((s) => s.missingLiveResolverFields)
      .filter(Boolean)
      .flat()
  }
  return snapshot.missingLiveResolverFields
}

function getFragmentResult(
  cacheKey: string,
  snapshot: SingularOrPluralSnapshot,
  storeEpoch: number
): FragmentResult {
  if ('length' in snapshot) {
    return {
      cacheKey,
      snapshot,
      data: snapshot.map((s) => s.data),
      isMissingData: isMissingData(snapshot),
      storeEpoch,
    }
  }
  return {
    cacheKey,
    snapshot,
    data: snapshot.data,
    isMissingData: isMissingData(snapshot),
    storeEpoch,
  }
}

class ClientEdgeQueryResultsCache {
  #cache: Map<string, [Array<QueryResult>, SuspenseResource]> = new Map()
  #retainCounts: Map<string, number> = new Map()
  #environment: IEnvironment

  constructor(environment: IEnvironment) {
    this.#environment = environment
  }

  get(fragmentIdentifier: string): void | Array<QueryResult> {
    return this.#cache.get(fragmentIdentifier)?.[0] ?? undefined
  }

  recordQueryResults(
    fragmentIdentifier: string,
    value: QueryResult[] // may be mutated after being passed here
  ): void {
    const existing = this.#cache.get(fragmentIdentifier)
    if (!existing) {
      const suspenseResource = new SuspenseResource(() =>
        this.#retain(fragmentIdentifier)
      )
      this.#cache.set(fragmentIdentifier, [value, suspenseResource])
      suspenseResource.temporaryRetain(this.#environment)
    } else {
      const [existingResults, suspenseResource] = existing
      value.forEach((queryResult) => {
        existingResults.push(queryResult)
      })
      suspenseResource.temporaryRetain(this.#environment)
    }
  }

  #retain(id: string): { dispose: () => void } {
    const retainCount = (this.#retainCounts.get(id) ?? 0) + 1
    this.#retainCounts.set(id, retainCount)
    return {
      dispose: () => {
        const newRetainCount = (this.#retainCounts.get(id) ?? 0) - 1
        if (newRetainCount > 0) {
          this.#retainCounts.set(id, newRetainCount)
        } else {
          this.#retainCounts.delete(id)
          this.#cache.delete(id)
        }
      },
    }
  }
}

class FragmentResource {
  #environment: IEnvironment
  #cache: FragmentResourceCache = LRUCache.create(CACHE_CAPACITY)
  #clientEdgeQueryResultsCache: void | ClientEdgeQueryResultsCache

  constructor(environment: IEnvironment) {
    this.#environment = environment
    if (RelayFeatureFlags.ENABLE_CLIENT_EDGES) {
      this.#clientEdgeQueryResultsCache = new ClientEdgeQueryResultsCache(
        environment
      )
    }
  }

  read(
    fragmentNode: ReaderFragment,
    fragmentRef: unknown,
    componentDisplayName: string,
    fragmentKey?: string
  ): FragmentResult | Promise<void> {
    return this.readWithIdentifier(
      fragmentNode,
      fragmentRef,
      getFragmentIdentifier(fragmentNode, fragmentRef),
      componentDisplayName,
      fragmentKey
    )
  }

  readWithIdentifier(
    fragmentNode: ReaderFragment,
    fragmentRef: unknown,
    fragmentIdentifier: string,
    componentDisplayName: string,
    fragmentKey?: string
  ): FragmentResult | Promise<void> {
    const environment = this.#environment

    if (fragmentRef == null) {
      return {
        cacheKey: fragmentIdentifier,
        data: null,
        isMissingData: false,
        snapshot: null,
        storeEpoch: 0,
      }
    }

    const storeEpoch = environment.getStore().getEpoch()

    if (fragmentNode?.metadata?.plural) {
      invariant(
        Array.isArray(fragmentRef),
        `solid-relay: Expected fragment pointer${
          fragmentKey != null ? ` for key \`${fragmentKey}\`` : ''
        } for fragment \`${fragmentNode.name}\` to be ` +
          `an array, instead got \`${typeof fragmentRef}\`. Remove \`@relay(plural: true)\` ` +
          `from fragment \`${fragmentNode.name}\` to allow the prop to be an object.`
      )
      if (fragmentRef.length === 0) {
        return {
          cacheKey: fragmentIdentifier,
          data: CONSTANT_READONLY_EMPTY_ARRAY,
          isMissingData: false,
          snapshot: CONSTANT_READONLY_EMPTY_ARRAY,
          storeEpoch,
        }
      }
    }

    const cachedValue = this.#cache.get(fragmentIdentifier)
    if (cachedValue != null) {
      if (cachedValue.kind === 'pending' && isPromise(cachedValue.promise)) {
        return cachedValue.promise
      } else if (
        cachedValue.kind === 'done' &&
        cachedValue.result.snapshot &&
        !missingLiveResolverFields(cachedValue.result.snapshot)?.length
      ) {
        this.#throwOrLogErrorsInSnapshot(cachedValue.result.snapshot)

        return cachedValue.result
      }
    }

    const fragmentSelector = getSelector(fragmentNode, fragmentRef)
    invariant(
      fragmentSelector != null,
      `solid-relay: Expected to receive an object where \`...${fragmentNode.name}\` was spread, ` +
        `but the fragment reference was not found\`. This is most ` +
        `likely the result of:\n` +
        `- Forgetting to spread \`${fragmentNode.name}\` in \`${componentDisplayName}\`'s parent's fragment.\n` +
        `- Conditionally fetching \`${
          fragmentNode.name
        }\` but unconditionally passing ${
          fragmentKey == null
            ? 'a fragment reference'
            : `the \`${fragmentKey}\``
        } prop ` +
        `to \`${componentDisplayName}\`. If the parent fragment only fetches the fragment conditionally ` +
        `- with e.g. \`@include\`, \`@skip\`, or inside a \`... on SomeType { }\` ` +
        `spread  - then the fragment reference will not exist. ` +
        `In this case, pass \`null\` if the conditions for evaluating the ` +
        `fragment are not met (e.g. if the \`@include(if)\` value is false.)`
    )

    const snapshot =
      'selectors' in fragmentSelector
        ? fragmentSelector.selectors.map((s) => environment.lookup(s))
        : environment.lookup(fragmentSelector)

    const fragmentResult = getFragmentResult(
      fragmentIdentifier,
      snapshot,
      storeEpoch
    )
    if (!fragmentResult.isMissingData) {
      this.#throwOrLogErrorsInSnapshot(snapshot)

      return fragmentResult
    }

    let clientEdgeRequests: RequestDescriptor[] | null = null
    if (
      RelayFeatureFlags.ENABLE_CLIENT_EDGES &&
      fragmentNode.metadata?.hasClientEdges === true &&
      hasMissingClientEdges(snapshot)
    ) {
      clientEdgeRequests = []
      const queryResource = getQueryResourceForEnvironment(this.#environment)
      const queryResults: QueryResult[] = []
      for (const snap of 'length' in snapshot ? snapshot : [snapshot]) {
        for (const {
          request,
          clientEdgeDestinationID,
        } of snap.missingClientEdges ?? []) {
          const { queryResult, requestDescriptor } =
            this.#performClientEdgeQuery(
              queryResource,
              fragmentNode,
              fragmentRef,
              request,
              clientEdgeDestinationID
            )
          if (isPromise(queryResult)) return queryResult
          queryResults.push(queryResult)
          clientEdgeRequests?.push(requestDescriptor)
        }
      }
      invariant(
        this.#clientEdgeQueryResultsCache != null,
        'Client edge query result cache should exist when ENABLE_CLIENT_EDGES is on.'
      )
      this.#clientEdgeQueryResultsCache.recordQueryResults(
        fragmentIdentifier,
        queryResults
      )
    }
    let clientEdgePromises: Promise<void>[] = []
    if (RelayFeatureFlags.ENABLE_CLIENT_EDGES && clientEdgeRequests) {
      clientEdgePromises = clientEdgeRequests
        .map((request) =>
          getPromiseForActiveRequest(this.#environment, request)
        )
        .filter(Boolean)
    }

    const fragmentOwner =
      'selectors' in fragmentSelector
        ? fragmentSelector.selectors[0].owner
        : fragmentSelector.owner
    const parentQueryPromiseResult =
      this.#getAndSavePromiseForFragmentRequestInFlight(
        fragmentIdentifier,
        fragmentNode,
        fragmentOwner,
        fragmentResult
      ) as { promise: Promise<void> } | undefined // TODO: remove typecast
    const parentQueryPromiseResultPromise = parentQueryPromiseResult?.promise
    const missingResolverFieldPromises =
      missingLiveResolverFields(snapshot)?.map(({ liveStateID }) => {
        const store = environment.getStore()

        // LiveResolverStore is not typed yet in @types/relay-runtime
        return (
          store as unknown as {
            getLiveResolverPromise(id: DataID): Promise<void>
          }
        ).getLiveResolverPromise(liveStateID)
      }) ?? []

    if (
      clientEdgePromises.length ||
      missingResolverFieldPromises.length ||
      isPromise(parentQueryPromiseResultPromise)
    ) {
      let promises: Promise<void>[] = []
      if (clientEdgePromises.length > 0) {
        promises = promises.concat(clientEdgePromises)
      }
      if (missingResolverFieldPromises.length > 0) {
        promises = promises.concat(missingResolverFieldPromises)
      }

      if (promises.length > 0) {
        if (parentQueryPromiseResultPromise) {
          promises.push(parentQueryPromiseResultPromise)
        }
        return Promise.all(promises).then((v) => void v)
      }

      if (parentQueryPromiseResultPromise) {
        return parentQueryPromiseResultPromise
      }
    }

    this.#throwOrLogErrorsInSnapshot(snapshot)

    return getFragmentResult(fragmentIdentifier, snapshot, storeEpoch)
  }

  #performClientEdgeQuery(
    queryResource: QueryResource,
    fragmentNode: ReaderFragment,
    fragmentRef: unknown,
    request: ConcreteRequest,
    clientEdgeDestinationID: DataID
  ): {
    requestDescriptor: RequestDescriptor
    queryResult: QueryResult | Promise<void>
  } {
    const originalVariables = getVariablesFromFragment(
      fragmentNode,
      fragmentRef
    )
    const variables = {
      ...originalVariables,
      [ID_KEY]: clientEdgeDestinationID,
    }
    const operation = createOperationDescriptor(request, variables, {})
    const fetchObservable = __internal.fetchQuery(this.#environment, operation)
    const queryResult = queryResource.prepare(operation, fetchObservable)
    return { requestDescriptor: operation.request, queryResult }
  }

  #throwOrLogErrorsInSnapshot(snapshot: SingularOrPluralSnapshot) {
    if ('length' in snapshot) {
      snapshot.forEach((s) => {
        handlePotentialSnapshotErrors(
          this.#environment,
          s.missingRequiredFields,
          s.relayResolverErrors
        )
      })
    } else {
      handlePotentialSnapshotErrors(
        this.#environment,
        snapshot.missingRequiredFields,
        snapshot.relayResolverErrors
      )
    }
  }

  readSpec(
    fragmentNodes: { [k: string]: ReaderFragment },
    fragmentRefs: { [k: string]: unknown },
    componentDisplayName: string
  ): { [k: string]: FragmentResult } | Promise<void> {
    const result: { [k: string]: FragmentResult } = {}
    for (const key in fragmentNodes) {
      const v = this.read(
        fragmentNodes[key],
        fragmentRefs[key],
        componentDisplayName,
        key
      )
      if (isPromise(v)) return v
      result[key] = v
    }
    return result
  }

  subscribe(fragmentResult: FragmentResult, callback: () => void): Disposable {
    const environment = this.#environment
    const { cacheKey } = fragmentResult
    const renderedSnapshot = fragmentResult.snapshot
    if (!renderedSnapshot) {
      return {
        // eslint-disable-next-line @typescript-eslint/no-empty-function
        dispose() {},
      }
    }

    const [didMissUpdates, currentSnapshot] =
      this.checkMissedUpdates(fragmentResult)

    if (didMissUpdates) callback()

    const disposables: Disposable[] = []
    if ('length' in renderedSnapshot) {
      invariant(
        currentSnapshot && 'length' in currentSnapshot,
        'solid-relay: Expected snapshots to be plural. ' +
          "If you're seeing this, this is likely a bug in solid-relay."
      )
      currentSnapshot.forEach((snapshot, idx) => {
        disposables.push(
          environment.subscribe(snapshot, (latestSnapshot) => {
            const storeEpoch = environment.getStore().getEpoch()
            this.#updatePluralSnapshot(
              cacheKey,
              currentSnapshot,
              latestSnapshot,
              idx,
              storeEpoch
            )
            callback()
          })
        )
      })
    } else {
      invariant(
        currentSnapshot != null && !('length' in currentSnapshot),
        'solid-relay: Expected snapshot to be singular. ' +
          "If you're seeing this, this is likely a bug in solid-relay."
      )
      disposables.push(
        environment.subscribe(currentSnapshot, (latestSnapshot) => {
          const storeEpoch = environment.getStore().getEpoch()
          this.#cache.set(cacheKey, {
            kind: 'done',
            result: getFragmentResult(cacheKey, latestSnapshot, storeEpoch),
          })
          callback()
        })
      )
    }

    if (RelayFeatureFlags.ENABLE_CLIENT_EDGES) {
      const clientEdgeQueryResults =
        this.#clientEdgeQueryResultsCache?.get(cacheKey) ?? undefined
      if (clientEdgeQueryResults?.length) {
        const queryResource = getQueryResourceForEnvironment(environment)
        clientEdgeQueryResults.forEach((queryResult) => {
          disposables.push(queryResource.retain(queryResult))
        })
      }
    }

    return {
      dispose: () => {
        disposables.forEach((s) => s.dispose())
        this.#cache.delete(cacheKey)
      },
    }
  }

  subscribeSpec(
    fragmentResults: { [k: string]: FragmentResult },
    callback: () => void
  ): Disposable {
    const disposables = Object.keys(fragmentResults).map((key) =>
      this.subscribe(fragmentResults[key], callback)
    )
    return {
      dispose: () => {
        disposables.forEach((disposable) => {
          disposable.dispose()
        })
      },
    }
  }

  checkMissedUpdates(
    fragmentResult: FragmentResult
  ): [boolean /* were updates missed? */, SingularOrPluralSnapshot | null] {
    const environment = this.#environment
    const renderedSnapshot = fragmentResult.snapshot
    if (!renderedSnapshot) {
      return [false, null]
    }

    let storeEpoch = null
    // Bail out if the store hasn't been written since last read
    storeEpoch = environment.getStore().getEpoch()
    if (fragmentResult.storeEpoch === storeEpoch) {
      return [false, fragmentResult.snapshot]
    }

    const { cacheKey } = fragmentResult

    if ('length' in renderedSnapshot) {
      let didMissUpdates = false
      const currentSnapshots: Snapshot[] = []
      renderedSnapshot.forEach((snapshot, idx) => {
        let currentSnapshot: Snapshot = environment.lookup(snapshot.selector)
        const renderData = snapshot.data
        const currentData = currentSnapshot.data
        const updatedData = recycleNodesInto(renderData, currentData)
        if (updatedData !== renderData) {
          currentSnapshot = { ...currentSnapshot, data: updatedData }
          didMissUpdates = true
        }
        currentSnapshots[idx] = currentSnapshot
      })
      // Only update the cache when the data is changed to avoid
      // returning different `data` instances
      if (didMissUpdates) {
        this.#cache.set(cacheKey, {
          kind: 'done',
          result: getFragmentResult(cacheKey, currentSnapshots, storeEpoch),
        })
      }
      return [didMissUpdates, currentSnapshots]
    }
    const currentSnapshot = environment.lookup(renderedSnapshot.selector)
    const renderData = renderedSnapshot.data
    const currentData = currentSnapshot.data
    const updatedData = recycleNodesInto(renderData, currentData)
    const updatedCurrentSnapshot: Snapshot = {
      data: updatedData,
      isMissingData: currentSnapshot.isMissingData,
      missingClientEdges: currentSnapshot.missingClientEdges,
      missingLiveResolverFields: currentSnapshot.missingLiveResolverFields,
      seenRecords: currentSnapshot.seenRecords,
      selector: currentSnapshot.selector,
      missingRequiredFields: currentSnapshot.missingRequiredFields,
      relayResolverErrors: currentSnapshot.relayResolverErrors,
    }
    if (updatedData !== renderData) {
      this.#cache.set(cacheKey, {
        kind: 'done',
        result: getFragmentResult(cacheKey, updatedCurrentSnapshot, storeEpoch),
      })
    }
    return [updatedData !== renderData, updatedCurrentSnapshot]
  }

  checkMissedUpdatesSpec(fragmentResults: {
    [k: string]: FragmentResult
  }): boolean {
    return Object.keys(fragmentResults).some(
      (key) => this.checkMissedUpdates(fragmentResults[key])[0]
    )
  }

  #getAndSavePromiseForFragmentRequestInFlight(
    cacheKey: string,
    fragmentNode: ReaderFragment,
    fragmentOwner: RequestDescriptor,
    fragmentResult: FragmentResult
  ): {
    promise: Promise<void>
    pendingOperations: readonly RequestDescriptor[]
  } | null {
    const pendingOperationsResult = getPendingOperationsForFragment(
      this.#environment,
      fragmentNode,
      fragmentOwner
    )
    if (pendingOperationsResult == null) {
      return null
    }

    // When the Promise for the request resolves, we need to make sure to
    // update the cache with the latest data available in the store before
    // resolving the Promise
    const networkPromise = pendingOperationsResult.promise
    const pendingOperations = pendingOperationsResult.pendingOperations
    const promise = networkPromise
      .then(() => {
        this.#cache.delete(cacheKey)
      })
      .catch<void>((error: Error) => {
        this.#cache.delete(cacheKey)
      })
    // @ts-expect-error(2339) Expando to annotate Promises.
    promise.displayName = (
      networkPromise as unknown as { displayName: string }
    ).displayName
    this.#cache.set(cacheKey, {
      kind: 'pending',
      pendingOperations,
      promise,
      result: fragmentResult,
    })
    return { promise, pendingOperations }
  }

  #updatePluralSnapshot(
    cacheKey: string,
    baseSnapshots: readonly Snapshot[],
    latestSnapshot: Snapshot,
    idx: number,
    storeEpoch: number
  ): void {
    const currentFragmentResult = this.#cache.get(cacheKey)
    if (isPromise(currentFragmentResult)) {
      reportInvalidCachedData(latestSnapshot.selector.node.name)
      return
    }

    const currentSnapshot = currentFragmentResult?.result?.snapshot
    if (currentSnapshot && !('length' in currentSnapshot)) {
      reportInvalidCachedData(latestSnapshot.selector.node.name)
      return
    }

    const nextSnapshots = currentSnapshot
      ? [...currentSnapshot]
      : [...baseSnapshots]
    nextSnapshots[idx] = latestSnapshot
    this.#cache.set(cacheKey, {
      kind: 'done',
      result: getFragmentResult(cacheKey, nextSnapshots, storeEpoch),
    })
  }
}

function reportInvalidCachedData(nodeName: string): void {
  invariant(
    false,
    `Relay: Expected to find cached data for plural fragment \`${nodeName}\` when ` +
      'receiving a subscription. ' +
      "If you're seeing this, this is likely a bug in Relay."
  )
}

export function createFragmentResource(
  environment: IEnvironment
): FragmentResource {
  return new FragmentResource(environment)
}

const dataResources: IMap<IEnvironment, FragmentResource> = WEAKMAP_SUPPORTED
  ? new WeakMap()
  : new Map()

export function getFragmentResourceForEnvironment(
  environment: IEnvironment
): FragmentResource {
  const cached = dataResources.get(environment)
  if (cached) return cached
  const newDataResource = createFragmentResource(environment)
  dataResources.set(environment, newDataResource)
  return newDataResource
}
