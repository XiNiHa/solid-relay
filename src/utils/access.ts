export type MaybeAccessor<T> = T | (() => T);
// biome-ignore lint/suspicious/noExplicitAny: extends
export type MaybeAccessorValue<T extends MaybeAccessor<any>> =
	// biome-ignore lint/suspicious/noExplicitAny: extends
	T extends () => any ? ReturnType<T> : T;

// biome-ignore lint/suspicious/noExplicitAny: extends
export const access = <T extends MaybeAccessor<any>>(
	v: T,
): MaybeAccessorValue<T> =>
	typeof v === "function" && !v.length ? v() : (v as MaybeAccessorValue<T>);
