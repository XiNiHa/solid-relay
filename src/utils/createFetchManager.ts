import type { Subscription } from 'relay-runtime'
import { onCleanup } from 'solid-js'

export function createFetchManager() {
  let subscription: Subscription | null = null
  let isFetching = false

  const manager = {
    get isFetching() {
      return isFetching
    },

    dispose: () => {
      if (subscription != null) {
        subscription.unsubscribe()
        subscription = null
      }
      isFetching = false
    },

    start: (newSubscription: Subscription) => {
      subscription = newSubscription
      isFetching = true
    },

    complete: () => {
      subscription = null
      isFetching = false
    },
  }

  onCleanup(() => manager.dispose())

  return manager
}
