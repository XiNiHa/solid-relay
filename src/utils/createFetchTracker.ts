import type { Subscription } from "relay-runtime";
import { createSignal, onCleanup } from "solid-js";

export function createFetchTracker(): {
	disposeFetch: () => void;
	startFetch: (subscription: Subscription) => void;
	completeFetch: () => void;
	isFetching: () => boolean;
} {
	let currentSubscription: Subscription | null = null;
	const [isFetching, setIsFetching] = createSignal(false);

	const disposeFetch = () => {
		if (currentSubscription != null) {
			currentSubscription.unsubscribe();
			currentSubscription = null;
		}
		setIsFetching(false);
	};

	const startFetch = (subscription: Subscription) => {
		currentSubscription = subscription;
		setIsFetching(true);
	};

	const completeFetch = () => {
		currentSubscription = null;
		setIsFetching(false);
	};

	onCleanup(disposeFetch);

	return {
		disposeFetch,
		startFetch,
		completeFetch,
		isFetching,
	};
}
