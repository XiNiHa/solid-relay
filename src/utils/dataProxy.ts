export type DataProxy<T> =
	| {
			(): T;
			readonly latest: T;
			readonly error: undefined;
			readonly pending: false;
	  }
	| {
			(): undefined;
			readonly latest: undefined;
			readonly error: unknown;
			readonly pending: false;
	  }
	| {
			(): undefined;
			readonly latest: undefined;
			readonly error: undefined;
			readonly pending: true;
	  };

export const makeDataProxy = <
	T extends {
		readonly data: unknown;
		readonly error: unknown;
		readonly pending: boolean;
	},
>(
	store: T,
	resource: () => void,
	data?: () => T["data"],
): DataProxy<T> => {
	const readData = () => {
		void resource();
		return data ? data() : Reflect.get(store, "data");
	};

	Object.defineProperties(readData, {
		latest: {
			get: () => Reflect.get(store, "data"),
		},
		error: {
			get: () => Reflect.get(store, "error"),
		},
		pending: {
			get: () => Reflect.get(store, "pending"),
		},
	});

	return readData as unknown as DataProxy<T>;
};
