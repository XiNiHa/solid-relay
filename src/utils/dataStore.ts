import { type Resource, untrack } from "solid-js";
import { type SetStoreFunction, createStore } from "solid-js/store";
import { useDataStores } from "../RelayEnvironment";
import type { MaybeAccessor } from "./access";

export type DataStore<T> =
	| {
			(): T;
			readonly latest: T;
			readonly error: undefined;
			readonly pending: false;
	  }
	| {
			(): undefined;
			readonly latest: undefined;
			readonly error: unknown;
			readonly pending: false;
	  }
	| {
			(): undefined;
			readonly latest: undefined;
			readonly error: undefined;
			readonly pending: true;
	  };

export const createDataStore = <
	T extends {
		readonly data: unknown;
		readonly error: unknown;
		readonly pending: boolean;
	},
>(
	init: T,
	maybeResource: MaybeAccessor<Resource<unknown> | undefined>,
): [DataStore<T>, SetStoreFunction<T>] => {
	const [store, setStore] = createStableStore(
		init,
		untrack(() => extractResource(maybeResource)),
	);

	const readData = () => {
		const resource = extractResource(maybeResource);
		void resource?.();
		const error = Reflect.get(store, "error");
		if (error) throw error;
		return Reflect.get(store, "data");
	};

	Object.defineProperties(readData, {
		latest: {
			get: () => Reflect.get(store, "data"),
		},
		error: {
			get: () => Reflect.get(store, "error"),
		},
		pending: {
			get: () => Reflect.get(store, "pending"),
		},
	});

	return [readData as unknown as DataStore<T>, setStore];
};

function createStableStore<
	T extends {
		readonly data: unknown;
		readonly error: unknown;
		readonly pending: boolean;
	},
>(init: T, resource: Resource<unknown> | undefined) {
	const stores = useDataStores();
	if (resource) {
		const existing = stores?.get(resource);
		if (existing) return existing as [T, SetStoreFunction<T>];
	}
	const store = createStore(init);
	if (resource) stores?.set(resource, store);
	return store;
}

function extractResource<T>(
	maybeResource: MaybeAccessor<Resource<T> | undefined>,
) {
	return maybeResource
		? "loading" in maybeResource
			? maybeResource
			: maybeResource()
		: undefined;
}
