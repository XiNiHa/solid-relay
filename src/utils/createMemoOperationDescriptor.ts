import { dequal } from "dequal/lite";
import {
	type CacheConfig,
	type GraphQLTaggedNode,
	type OperationDescriptor,
	type Variables,
	createOperationDescriptor,
	getRequest,
} from "relay-runtime";
import { createMemo } from "solid-js";
import type { Accessor } from "solid-js";
import { type MaybeAccessor, access } from "./access";

export function createMemoOperationDescriptor(
	gqlQuery: MaybeAccessor<GraphQLTaggedNode>,
	variables: MaybeAccessor<Variables | undefined>,
	cacheConfig?: MaybeAccessor<CacheConfig | undefined>,
): Accessor<OperationDescriptor | undefined> {
	const memoizedVariables = createMemo(() => access(variables), undefined, {
		equals: dequal,
	});
	const memoizedCacheConfig = createMemo(() => access(cacheConfig), undefined, {
		equals: dequal,
	});

	return createMemo(() => {
		const variables = memoizedVariables();
		if (!variables) return;

		return createOperationDescriptor(
			getRequest(access(gqlQuery)),
			variables,
			memoizedCacheConfig(),
		);
	});
}
