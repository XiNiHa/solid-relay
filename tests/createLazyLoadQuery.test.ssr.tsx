import {
	type ConcreteRequest,
	type GraphQLTaggedNode,
	RecordSource,
	Store,
	graphql,
} from "relay-runtime";
import { type MockEnvironment, createMockEnvironment } from "relay-test-utils";
import { ErrorBoundary, type JSXElement, Suspense } from "solid-js";
import { RelayEnvironmentProvider, createLazyLoadQuery } from "solid-relay";
import type { createLazyLoadQuerySSRTestQuery } from "./__generated__/createLazyLoadQuerySSRTestQuery.graphql";
import { renderStream } from "./utils";

let environment: MockEnvironment;
const View = (props: { children: JSXElement }) => (
	<RelayEnvironmentProvider environment={environment}>
		{props.children}
	</RelayEnvironmentProvider>
);

describe("createLazyLoadQuery SSR", () => {
	const gqlQuery = graphql`
		query createLazyLoadQuerySSRTestQuery {
			node(id: "1") {
				id
				... on User {
					name
				}
			}
		}
	` as GraphQLTaggedNode & ConcreteRequest;
	const Comp = (props: {
		gqlQuery: GraphQLTaggedNode;
	}) => {
		const data = createLazyLoadQuery<createLazyLoadQuerySSRTestQuery>(
			props.gqlQuery,
			{},
		);
		return (
			<ErrorBoundary
				fallback={(err) => <h1 data-testid="error">{err.message}</h1>}
			>
				<Suspense fallback="Fallback">
					<h1 data-testid="name">{data()?.node?.name}</h1>
				</Suspense>
			</ErrorBoundary>
		);
	};

	beforeEach(() => {
		environment = createMockEnvironment({
			store: new Store(new RecordSource(), { gcReleaseBufferSize: 0 }),
		});
	});

	it("renders", async () => {
		const { shellCompleted, readable } = renderStream(() => (
			<View>
				<Comp gqlQuery={gqlQuery} />
			</View>
		));
		await shellCompleted;
		environment.mock.resolve(gqlQuery, {
			data: { node: { __typename: "User", id: "1", name: "Alice" } },
		});
		expect(await Array.fromAsync(readable)).toMatchSnapshot();
	});
});
