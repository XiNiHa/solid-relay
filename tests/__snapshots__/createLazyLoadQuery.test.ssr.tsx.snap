// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createLazyLoadQuery SSR > renders 1`] = `
[
  "<!--!$e000010--><template id="pl-0000100"></template>Fallback<!--pl-0000100--><!--!$/e000010--><script>self.$R=self.$R||[];_$HY.r["00000"]=$R[0]=($R[2]=r=>(r.p=new Promise((s,f)=>{r.s=s,r.f=f})))($R[1]={p:0,s:0,f:0});_$HY.r["0000100"]=$R[3]=$R[2]($R[4]={p:0,s:0,f:0});</script>",
  "<template id="0000100"><h1 data-hk=00001000 data-testid="name">Alice</h1></template>",
  "<script>$df("0000100");function $df(e,n,o,t){if(n=document.getElementById(e),o=document.getElementById("pl-"+e)){for(;o&&8!==o.nodeType&&o.nodeValue!=="pl-"+e;)t=o.nextSibling,o.remove(),o=t;_$HY.done?o.remove():o.replaceWith(n.content)}n.remove(),_$HY.fe(e)};($R[9]=(r,d)=>{r.s(d),r.p.s=1,r.p.v=d})($R[1],$R[5]=($R[6]=d=>new ReadableStream({start:c=>{d.on({next:v=>{c.enqueue(v)},throw:v=>{c.error(v)},return:()=>{c.close()}})}}))($R[7]=($R[8]=(b,a,s,l,p,f,e,n)=>(b=[],a=!0,s=!1,l=[],p=0,f=(v,m,x)=>{for(x=0;x<p;x++)l[x]&&l[x][m](v)},n=(o,x,z,c)=>{for(x=0,z=b.length;x<z;x++)(c=b[x],(!a&&x===z-1)?o[s?"return":"throw"](c):o.next(c))},e=(o,t)=>(a&&(l[t=p++]=o),n(o),()=>{a&&(l[t]=void 0)}),{__SEROVAL_STREAM__:!0,on:o=>e(o),next:v=>{a&&(b.push(v),f(v,"next"))},throw:v=>{a&&(b.push(v),f(v,"throw"),a=s=!1,l.length=0)},return:v=>{a&&(b.push(v),f(v,"return"),a=!1,s=!0,l.length=0)}}))()));$R[9]($R[4],!0);$R[7].next($R[10]={data:$R[11]={node:$R[12]={__typename:"User",id:"1",name:"Alice"}}});$R[7].return(void 0);</script>",
]
`;
